File: src/main/java/me/testaccount666/serversystem/clickablesigns/executables/warp/ConfiguratorWarpSign.java
package me.testaccount666.serversystem.clickablesigns.executables.warp;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator;
import me.testaccount666.serversystem.clickablesigns.SignType;
import me.testaccount666.serversystem.commands.executables.warp.manager.WarpManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.ChatColor;
import me.testaccount666.serversystem.utils.ComponentColor;
import org.bukkit.block.Sign;
import org.bukkit.block.sign.Side;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import static me.testaccount666.serversystem.utils.MessageBuilder.sign;

public class ConfiguratorWarpSign extends AbstractSignConfigurator {

    @Override
    protected String getCreatePermissionNode() {
        return "ClickableSigns.Warp.Create";
    }

    @Override
    protected SignType getSignType() {
        return SignType.WARP;
    }

    @Override
    protected boolean validateConfiguration(User user, Sign sign, YamlConfiguration config) {
        var warpManager = ServerSystem.getInstance().getRegistry().getService(WarpManager.class);
        if (warpManager == null) {
            sign("Warp.NoWarpManager", user).build();
            return false;
        }

        var front = sign.getSide(Side.FRONT);
        var warpName = front.getLine(1);
        if (warpName.isEmpty()) {
            sign("Warp.NoWarpSpecified", user).build();
            return false;
        }

        if (!warpManager.warpExists(warpName.toLowerCase())) {
            sign("Warp.WarpNotFound", user)
                    .postModifier(message -> message.replace("<WARP>", warpName)).build();
            return false;
        }

        front.line(0, ComponentColor.translateToComponent(SignType.WARP.signName()));
        front.line(1, ComponentColor.translateToComponent("&2${warpName}"));
        var back = sign.getSide(Side.BACK);
        for (var index = 0; index < 4; index++) back.line(index, front.line(index));
        sign.update();
        return true;
    }

    @Override
    protected void addSignSpecificConfiguration(User user, Sign sign, FileConfiguration config) {
        var warpName = sign.getSide(Side.FRONT).getLine(1);
        warpName = ChatColor.stripColor(warpName);
        config.set("WarpName", warpName);
    }

    @Override
    protected String getSuccessMessageKey() {
        return "Warp.Created";
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/executables/warp/ConfiguratorWarpSign.kt
package me.testaccount666.serversystem.clickablesigns.executables.warp

import me.testaccount666.paperktx.colors.ChatColor.Companion.stripColor
import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator
import me.testaccount666.serversystem.clickablesigns.SignType
import me.testaccount666.serversystem.commands.executables.waypoints.warp.manager.WarpManager
import me.testaccount666.serversystem.extensions.getServiceOrNull
import me.testaccount666.serversystem.extensions.signMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.block.Sign
import org.bukkit.block.sign.Side
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration

class ConfiguratorWarpSign : AbstractSignConfigurator() {
    override val createPermissionNode = "ClickableSigns.Warp.Create"
    override val signType = SignType.WARP
    override val successMessageKey = "Warp.Created"


    override fun validateConfiguration(user: User, sign: Sign, config: YamlConfiguration): Boolean {
        val warpManager = getServiceOrNull<WarpManager>() ?: run {
            user.signMsg("Warp.NoWarpManager")
            return false
        }

        val front = sign.getSide(Side.FRONT)
        val warpName = front.getLine(1)
        if (warpName.isEmpty()) {
            user.signMsg("Warp.NoWarpSpecified")
            return false
        }

        if (!warpManager.pointExists(warpName)) {
            user.signMsg("Warp.WarpNotFound") {
                postModifier { it.replace("<WARP>", warpName) }
            }
            return false
        }

        front.line(0, SignType.WARP.signName.asComponent())
        front.line(1, "&2${warpName}".asComponent())
        val back = sign.getSide(Side.BACK)
        for (index in 0..3) back.line(index, front.line(index))
        sign.update()
        return true
    }

    override fun addSignSpecificConfiguration(user: User, sign: Sign, config: FileConfiguration) {
        var warpName = sign.getSide(Side.FRONT).getLine(1)
        warpName = stripColor(warpName)
        config["WarpName"] = warpName
    }
}